docs: Add receive command docs (#2726)
thanos-io/thanos: docs/components/receive.md

---
title: Receiver
type: docs
menu: components
---

# Receiver

The `thanos receive` command implements the [Prometheus Remote Write API](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_write). It builds on top of existing Prometheus TSDB and retains their usefulness while extending their functionality with long-term-storage, horizontal scalability, and downsampling. It exposes the StoreAPI so that [Thanos Queriers](./query.md) can query received metrics in real-time. The [Thanos Sidecar](./sidecar.md) is not sufficient for this, as the system would always lag the block length behind (typically 2 hours).

> We recommend this component to users who can only push into a Thanos due to air-gapped environments. Please note the various pros and cons of pushing metrics.

Reviewer: ```suggestion
We recommend this component to users who can only push into a Thanos due to air-gapped, or egress only environments. Please note the various pros and cons of pushing metrics.
```

Otherwise well put :muscle:

> Thanos Receive supports multi-tenancy by using labels. We don't recommend the Thanos Receiver in order to achieve a global view of data of a single-tenant. For such use cases, the [Thanos Sidecar](./sidecar.md) based approach with layered [Thanos Queriers](./query.md) is recommended. Note that, multi-tenancy may also be achievable if ingestion is not user-controlled, as then enforcing of labels, for example using the [prom-label-proxy](https://github.com/openshift/prom-label-proxy) (please thoroughly understand the mechanism if intending to employ this mechanism, as the wrong configuration could leak data).

Reviewer: This kind of suggest again that you can achieve multitenancy with receiver only, which is not true. Let's maybe just save this section to some `Multitenancy` Operating doc? :thinking: I would here only say:
```suggestion
Thanos Receive supports multi-tenancy by using labels.  See [Multitenancy documentation here](link).
```

Author: I have started a doc and linked. FYI it's just bare bones. I can send an another PR to improve it.

---
title: Receiver
type: docs
menu: components
---

# Receiver

> The `thanos receive` command implements the [Prometheus Remote Write API](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_write). It builds on top of existing Prometheus servers and retains their usefulness while extending their functionality with long-term-storage, horizontal scalability, and downsampling. The [Thanos Sidecar](./sidecar.md) is not sufficient for this, as the system would always lag the block length behind (typically 2 hours), which would prevent the most common query pattern of Prometheus: real-time queries of very recent data.

Reviewer 2: I don't think the last sentence describes this particularly good.
I'm not even fully grasping what you're trying to say.

Author: I've cheated and copied that over proposal document 😅 Happy to remove or paraphrase it.

Reviewer 3: Hmm for the sentence starting with `It builds on top of existing Prometheus servers...` I think is a little bit fuzzy for people new to Thanos and who aren't familiar with everything. Also if Thanos builds on top of Prometheus. Maybe we should collaborate on a Google doc so we can iterate faster. Wdyt?

Author: @squat Yes, please. Greate idea. https://docs.google.com/document/d/1H4psm3ancNRyUfg4UilroAEKb96zm_dldH6UnVoLB6A/edit#

> This component is only recommended for uses for whom pushing is the only viable solution, for example, analytics use cases or cases where the data ingestion must be client-initiated, such as software as a service type environments.

Reviewer 2: ```suggestion
We recommend this component to users who can only push into a Thanos due to air-gapped environments. Examples are: analytic use cases or cases where the data ingestion must be client-initiated, such as software as a service type environments.
```

I don't see why the receiver is particularily good for analytic use case compared to the sidecar.

Reviewer: Agree

Reviewer: At least for now, there is no difference for analytics

> This component is not recommended in order to achieve a global view of data of a single-tenant, for those cases the sidecar based approach with layered Thanos Queriers is recommended. Multi-tenancy may also be achievable if ingestion is not user-controlled, as then enforcing of labels, for example using the [prom-label-proxy](https://github.com/openshift/prom-label-proxy) (please thoroughly understand the mechanism if intending to employ this mechanism, as the wrong configuration could leak data).

Reviewer 2: ```suggestion
We don't recommend the Thanos Receiver in order to achieve a global view of data of a single-tenant. For such use cases, the sidecar based approach with layered Thanos Queriers is recommended. Multi-tenancy may also be achievable if ingestion is not user-controlled, as then enforcing of labels, for example using the [prom-label-proxy](https://github.com/openshift/prom-label-proxy) (please thoroughly understand the mechanism if intending to employ this mechanism, as the wrong configuration could leak data).
```

We we really need to touch on multi-tenancy here?

Author: We don't I can remove it. From the proposal as well.

Reviewer: I think it is awesome Kemal touched multitenancy, but I would elaborate more on this part... This sentence might be confusing

> Also, users are asked to note the [various pros and cons of pushing metrics](https://docs.google.com/document/d/1H47v7WfyKkSLMrR8_iku6u9VB73WrVzBHb2SB6dL9_g/edit#heading=h.2v27snv0lsur).

Reviewer 2: ```suggestion
Additionally, we would like to ask users to note the [various pros and cons of pushing metrics](https://docs.google.com/document/d/1H47v7WfyKkSLMrR8_iku6u9VB73WrVzBHb2SB6dL9_g/edit#heading=h.2v27snv0lsur).
```

Reviewer 2: Even better is probably something like:
> Please note the various pros and cons of pushing metrics.
